[Bug] Admission controller crushes with error: index out of range [1] with length 1

### Kyverno Version

1.11.5

### Kubernetes Version

1.25.x

### Kubernetes Platform

OpenShift

### Kyverno Rule Type

Generate

### Description

We have another policy to work with VPA, based on `Deployment/scale` kind
Create policy should **generate** VPA if replicas count != 0 and there's no VPA objects without `canremove: true` label
Delete policy should **mutate** VPA if replicas count == 0
The problem is that these policies don't work together, only individually, if I apply manifest with create and delete policies together the admission-controller will crush

I'm tried to troubleshoot policy as described in this [guide](https://kyverno.io/docs/troubleshooting/), but it won't help.

Info about cluster:

```
Server Version: 4.12.0-0.okd-2023-02-04-212953
Kubernetes Version: v1.25.0-2653+a34b9e9499e6c3-dirty
```

### Steps to reproduce

1. Install [VPA](https://github.com/kubernetes/autoscaler/tree/master/vertical-pod-autoscaler)
2. Apply Kyverno policy manifest:
```yaml
---
apiVersion: kyverno.io/v1
kind: ClusterPolicy
metadata:
  name: create-delete-vpa-on-replicas
spec:
  background: false
  rules:
    - name: create-vpa-on-non-zero-replicas
      match:
        all:
          - resources:
              kinds:
                - Deployment/scale
      exclude:
        any:
          - resources:
              namespaces:
                - "devops-*"
                - "openshift-*"
                - "kube-*"
                - "calico-*"
                - "argocd-*"
                - "tigera-operator"
      context:
        - name: randomuidcreate
          variable:
            jmesPath: random('[a-z0-9]{5}')
        - name: checkreplicascreate
          variable:
            jmesPath: "request.object.spec.replicas || '0'"
        - name: GenerateVPACheckExistence
          apiCall:
            urlPath: "/apis/autoscaling.k8s.io/v1/namespaces/{{request.namespace}}/verticalpodautoscalers"
            jmesPath: "length(items[].metadata.labels | [? canremove!='true'])"
      preconditions:
        all:
          - key: "{{ checkreplicascreate }}"
            operator: NotEquals
            value: "0"
          - key: "{{ GenerateVPACheckExistence }}"
            operator: Equals
            value: 0
      generate:
        apiVersion: autoscaling.k8s.io/v1
        kind: VerticalPodAutoscaler
        name: "{{request.object.metadata.name}}-deployment-{{randomuidcreate}}"
        namespace: "{{request.object.metadata.namespace}}"
        synchronize: false
        data:
          spec:
            targetRef:
              apiVersion: "apps/v1"
              kind: "Deployment"
              name: "{{request.object.metadata.name}}"
            updatePolicy:
              updateMode: "Off"
            resourcePolicy:
              containerPolicies:
                - containerName: "*"
    - name: delete-vpa-on-zero-replicas
      match:
        all:
          - resources:
              kinds:
                - Deployment/scale
      exclude:
        any:
          - resources:
              namespaces:
                - "devops-*"
                - "openshift-*"
                - "kube-*"
                - "calico-*"
                - "argocd-*"
                - "tigera-operator"
      context:
        - name: checkreplicasdelete
          variable:
            jmesPath: "request.object.spec.replicas || '0'"
      preconditions:
        all:
          - key: "{{ checkreplicasdelete }}"
            operator: Equals
            value: "0"
      mutate:
        targets:
          - apiVersion: autoscaling.k8s.io/v1
            kind: VerticalPodAutoscaler
            namespace: "{{request.object.metadata.namespace}}"
            name: "{{request.object.metadata.name}}-deployment-*"
        patchesJson6902: |-
          - path: "/metadata/labels"
            op: remove
          - path: "/metadata/labels/canremove"
            op: add
            value: "true"
```
3. Apply Deployment manifest:
```yaml
kind: Deployment
apiVersion: apps/v1
metadata:
  name: test
  namespace: test
  labels:
    app: test
spec:
  replicas: 0
  selector:
    matchLabels:
      app: test
  template:
    metadata:
      labels:
        app: test
    spec:
      containers:
        - name: container
          image: nginxinc/nginx-unprivileged:latest
  strategy:
    type: Recreate
```
4. Run scale to replicas 1 command:
```bash
oc scale deployment test --replicas=1
```
After that the admission-controller should be crushed
5. Run scale to replicas 0 command:
```bash
oc scale deployment test --replicas=0
```
After that the admission-controller should be crushed again

### Expected behavior

Successfully generation/mutation of VPA object

### Screenshots

Proof that this is the only policy that applied
![image](https://github.com/user-attachments/assets/ca4a828d-b53a-4d37-b924-ed1f1e2adbef)


### Kyverno logs

```Shell
Admission-controller

On create policy:

	github.com/kyverno/kyverno/pkg/webhooks/resource/updaterequest.go:23 +0x23b
created by github.com/kyverno/kyverno/pkg/webhooks/resource.(*resourceHandlers).handleBackgroundApplies in goroutine 194569
	github.com/kyverno/kyverno/pkg/webhooks/resource/updaterequest.go:54 +0x3bd
github.com/kyverno/kyverno/pkg/webhooks/resource.(*resourceHandlers).handleMutateExisting(_, {_, _}, {{_, _}, _}, {{0xc002c1c810, 0x24}, {{0xc001fad860, 0xb}, ...}, ...}, ...)
	github.com/kyverno/kyverno/pkg/engine/engine.go:164 +0x185
github.com/kyverno/kyverno/pkg/engine.(*engine).ApplyBackgroundChecks(0xc000c0a420, {0x411b860, 0xc0025f6ff0}, {0x4142fc0, 0xc001666b40})
	github.com/kyverno/kyverno/pkg/engine/background.go:26 +0x9a
github.com/kyverno/kyverno/pkg/engine.(*engine).applyBackgroundChecks(0x41268b0?, {0xc001abb880?, 0xc0029fffc0?}, {{0x41268b0?, 0xc001abb880?}, 0x4137788?}, {0x4142fc0, 0xc001666b40})
	github.com/kyverno/kyverno/pkg/engine/background.go:39 +0x1f9
github.com/kyverno/kyverno/pkg/engine.(*engine).filterRules(0xc001daaa00?, {0x4142fc0, 0xc001666b40}, {{0x41268b0?, 0xc001abb880?}, 0xc0025f7fb0?}, {0xc001e3c840?, 0x4160db8?, 0xc001abb880?})
	github.com/kyverno/kyverno/pkg/engine/background.go:104 +0xbd0
github.com/kyverno/kyverno/pkg/engine.(*engine).filterRule(_, {{0xc0025ecfa0, 0x1f}, {0xc0012245a0, 0x3, 0x3}, {{0x0, 0x0, 0x0}, {0xc001aae820, ...}, ...}, ...}, ...)
	github.com/kyverno/kyverno/pkg/engine/variables/evaluate.go:34 +0xaa
github.com/kyverno/kyverno/pkg/engine/variables.EvaluateConditions({{0x41268b0?, 0xc001abb8c0?}, 0xc0012245a0?}, {0x7fc479fc7830?, 0xc0002ebdc0?}, {0x367fba0?, 0xc001e085a0?})
	github.com/kyverno/kyverno/pkg/engine/variables/evaluate.go:85 +0x217
github.com/kyverno/kyverno/pkg/engine/variables.evaluateAnyAllConditions({{0x41268b0?, 0xc001abb8c0?}, 0xc001d9d9f8?}, {0x7fc479fc7830, 0xc0002ebdc0}, {{0x0, 0x0, 0x0}, {0xc001c24060, 0x2, ...}})
	github.com/kyverno/kyverno/pkg/engine/variables/evaluate.go:15 +0x85
github.com/kyverno/kyverno/pkg/engine/variables.Evaluate({{0x41268b0?, 0xc001abb8c0?}, 0xc000100000?}, {0x7fc479fc7830, 0xc0002ebdc0}, {0xc00196b2d8, {0xc001fbfbd8, 0x6}, 0xc00196b2f0, {0x0, ...}})
	github.com/kyverno/kyverno/pkg/engine/variables/vars.go:65 +0xe5
github.com/kyverno/kyverno/pkg/engine/variables.SubstituteAllInPreconditions({{0x41268b0?, 0xc001abb8c0?}, 0xc0022ff1c0?}, {0x7fc479fc7830, 0xc0002ebdc0}, {0x330a800?, 0xc0022ff1e0?})
	github.com/kyverno/kyverno/pkg/engine/variables/vars.go:181 +0x9e
github.com/kyverno/kyverno/pkg/engine/variables.substituteAll({{0x41268b0?, 0xc001abb8c0?}, 0xc000f27438?}, {0x7fc479fc7830, 0xc0002ebdc0}, {0x330a800?, 0xc0022ff200?}, 0x4108c5?)
	github.com/kyverno/kyverno/pkg/engine/jsonutils/traverse.go:33 +0xef
github.com/kyverno/kyverno/pkg/engine/variables.substituteVars({{0x41268b0?, 0xc001abb8c0?}, 0x7fc5c11993c8?}, {0x7fc479fc7830?, 0xc0002ebdc0?}, {0x330a800?, 0xc0022ff220}, 0xc000f273c0?)
	github.com/kyverno/kyverno/pkg/engine/jsonutils/traverse.go:61
github.com/kyverno/kyverno/pkg/engine/jsonutils.(*Traversal).TraverseJSON(...)
github.com/kyverno/kyverno/pkg/engine/jsonutils.(*Traversal).traverseJSON(0xc000f27338, {0x330a800?, 0xc0022ff220}, {0x0, 0x0})
	github.com/kyverno/kyverno/pkg/engine/jsonutils/traverse.go:66 +0xf0
	github.com/kyverno/kyverno/pkg/engine/variables/vars.go:215 +0x85
github.com/kyverno/kyverno/pkg/engine/variables.substituteVariablesIfAny.OnlyForLeafsAndKeys.func2(0x28?)
	github.com/kyverno/kyverno/pkg/engine/variables/vars.go:44 +0x5a
github.com/kyverno/kyverno/pkg/engine/variables.substituteVariablesIfAny.func1(0xc0027b92c0)
	github.com/kyverno/kyverno/pkg/engine/variables/vars.go:323 +0x2c
github.com/kyverno/kyverno/pkg/engine/variables.SubstituteAllInPreconditions.newPreconditionsVariableResolver.func1({0x7fc479fc7830?, 0xc0002ebdc0?}, {0xc0021c8121, 0x19})
	github.com/kyverno/kyverno/pkg/engine/variables/vars.go:365 +0x442
github.com/kyverno/kyverno/pkg/engine/variables.DefaultVariableResolver({0x7fc479fc7830?, 0xc0002ebdc0?}, {0xc0021c8121?, 0xc0003cc140?})
	github.com/kyverno/kyverno/pkg/engine/context/evaluate.go:13 +0x6b
github.com/kyverno/kyverno/pkg/engine/context.(*context).Query(0xc0002ebdc0, {0xc0021c8121, 0x19})
github.com/kyverno/kyverno/pkg/engine/context.(*deferredLoaders).match(0xc0025f7080, {0xc0021c8121, 0x19}, 0x2, 0x6)
github.com/kyverno/kyverno/pkg/engine/context.(*context).loadDeferred(...)
	github.com/kyverno/kyverno/pkg/engine/context/deferred.go:146 +0x69
	github.com/kyverno/kyverno/pkg/engine/context/deferred.go:172 +0x139
github.com/kyverno/kyverno/pkg/engine/context.(*deferredLoaders).LoadMatching(0xc0025f7080, {0xc0021c8121, 0x19}, 0x51a7be?)
	github.com/kyverno/kyverno/pkg/engine/context/evaluate.go:43
goroutine 194570 [running]:

panic: runtime error: index out of range [1] with length 1
```

On delete policy:
```
	github.com/kyverno/kyverno/pkg/webhooks/resource/updaterequest.go:23 +0x23b
created by github.com/kyverno/kyverno/pkg/webhooks/resource.(*resourceHandlers).handleBackgroundApplies in goroutine 194737
	github.com/kyverno/kyverno/pkg/webhooks/resource/updaterequest.go:54 +0x3bd
github.com/kyverno/kyverno/pkg/webhooks/resource.(*resourceHandlers).handleMutateExisting(_, {_, _}, {{_, _}, _}, {{0xc0020210b0, 0x24}, {{0xc0015374f0, 0xb}, ...}, ...}, ...)

	github.com/kyverno/kyverno/pkg/engine/engine.go:164 +0x185
github.com/kyverno/kyverno/pkg/engine.(*engine).ApplyBackgroundChecks(0xc001cd4d10, {0x411b860, 0xc002942c60}, {0x4142fc0, 0xc00127fa40})
	github.com/kyverno/kyverno/pkg/engine/background.go:26 +0x9a
github.com/kyverno/kyverno/pkg/engine.(*engine).applyBackgroundChecks(0x41268b0?, {0xc001b7ac40?, 0xc002471830?}, {{0x41268b0?, 0xc001b7ac40?}, 0x4137788?}, {0x4142fc0, 0xc00127fa40})
	github.com/kyverno/kyverno/pkg/engine/background.go:39 +0x1f9
github.com/kyverno/kyverno/pkg/engine.(*engine).filterRules(0xc0024f2440?, {0x4142fc0, 0xc00127fa40}, {{0x41268b0?, 0xc001b7ac40?}, 0xc001cfe210?}, {0xc001cfe360?, 0x4160db8?, 0xc001b7ac40?})
	github.com/kyverno/kyverno/pkg/engine/background.go:104 +0xbd0
github.com/kyverno/kyverno/pkg/engine.(*engine).filterRule(_, {{0xc001a87560, 0x1f}, {0xc002167560, 0x3, 0x3}, {{0x0, 0x0, 0x0}, {0xc003223790, ...}, ...}, ...}, ...)
	github.com/kyverno/kyverno/pkg/engine/variables/evaluate.go:34 +0xaa
github.com/kyverno/kyverno/pkg/engine/variables.EvaluateConditions({{0x41268b0?, 0xc001b7ac80?}, 0xc002167560?}, {0x7faa2c8937c8?, 0xc0005d8540?}, {0x367fba0?, 0xc001cff0b0?})
	github.com/kyverno/kyverno/pkg/engine/variables/evaluate.go:85 +0x217
github.com/kyverno/kyverno/pkg/engine/variables.evaluateAnyAllConditions({{0x41268b0?, 0xc001b7ac80?}, 0xc00398f710?}, {0x7faa2c8937c8, 0xc0005d8540}, {{0x0, 0x0, 0x0}, {0xc001403f80, 0x2, ...}})
	github.com/kyverno/kyverno/pkg/engine/variables/evaluate.go:15 +0x85
github.com/kyverno/kyverno/pkg/engine/variables.Evaluate({{0x41268b0?, 0xc001b7ac80?}, 0xc0000b9400?}, {0x7faa2c8937c8, 0xc0005d8540}, {0xc00398fa40, {0xc002250d00, 0x9}, 0xc00398fa58, {0x0, ...}})
	github.com/kyverno/kyverno/pkg/engine/variables/vars.go:65 +0xe5
github.com/kyverno/kyverno/pkg/engine/variables.SubstituteAllInPreconditions({{0x41268b0?, 0xc001b7ac80?}, 0x26?}, {0x7faa2c8937c8, 0xc0005d8540}, {0x330a800?, 0xc002471c40?})
	github.com/kyverno/kyverno/pkg/engine/variables/vars.go:181 +0x9e
github.com/kyverno/kyverno/pkg/engine/variables.substituteAll({{0x41268b0?, 0xc001b7ac80?}, 0xc001769438?}, {0x7faa2c8937c8, 0xc0005d8540}, {0x330a800?, 0xc002471c60?}, 0x367fba0?)
	github.com/kyverno/kyverno/pkg/engine/variables/vars.go:215 +0x85
github.com/kyverno/kyverno/pkg/engine/variables.substituteVars({{0x41268b0?, 0xc001b7ac80?}, 0x7fab73c80878?}, {0x7faa2c8937c8?, 0xc0005d8540?}, {0x330a800?, 0xc002471c80}, 0xc0017693c0?)
github.com/kyverno/kyverno/pkg/engine/variables.substituteVariablesIfAny.func1(0xc001cff4d0)
	github.com/kyverno/kyverno/pkg/engine/variables/vars.go:44 +0x5a
	github.com/kyverno/kyverno/pkg/engine/jsonutils/traverse.go:66 +0xf0
github.com/kyverno/kyverno/pkg/engine/jsonutils.(*Traversal).traverseJSON(0xc001769338, {0x330a800?, 0xc002471c80}, {0x0, 0x0})
	github.com/kyverno/kyverno/pkg/engine/jsonutils/traverse.go:33 +0xef
github.com/kyverno/kyverno/pkg/engine/variables.substituteVariablesIfAny.OnlyForLeafsAndKeys.func2(0xc000e60280?)
	github.com/kyverno/kyverno/pkg/engine/variables/vars.go:365 +0x442
github.com/kyverno/kyverno/pkg/engine/jsonutils.(*Traversal).TraverseJSON(...)
github.com/kyverno/kyverno/pkg/engine/variables.SubstituteAllInPreconditions.newPreconditionsVariableResolver.func1({0x7faa2c8937c8?, 0xc0005d8540?}, {0xc00388af31, 0x13})
	github.com/kyverno/kyverno/pkg/engine/jsonutils/traverse.go:61
	github.com/kyverno/kyverno/pkg/engine/variables/vars.go:323 +0x2c
goroutine 193010 [running]:
	github.com/kyverno/kyverno/pkg/engine/context/evaluate.go:13 +0x6b
github.com/kyverno/kyverno/pkg/engine/context.(*context).Query(0xc0005d8540, {0xc00388af31, 0x13})
panic: runtime error: index out of range [1] with length 1
github.com/kyverno/kyverno/pkg/engine/context.(*deferredLoaders).match(0xc002942d20, {0xc00388af31, 0x13}, 0x2, 0x6)
	github.com/kyverno/kyverno/pkg/engine/context/evaluate.go:43
github.com/kyverno/kyverno/pkg/engine/context.(*deferredLoaders).LoadMatching(0xc002942d20, {0xc00388af31, 0x13}, 0x51a7be?)
	github.com/kyverno/kyverno/pkg/engine/context/deferred.go:172 +0x139

github.com/kyverno/kyverno/pkg/engine/context.(*context).loadDeferred(...)
github.com/kyverno/kyverno/pkg/engine/variables.DefaultVariableResolver({0x7faa2c8937c8?, 0xc0005d8540?}, {0xc00388af31?, 0xc0003a6280?})
	github.com/kyverno/kyverno/pkg/engine/context/deferred.go:146 +0x69
```


### Slack discussion

_No response_

### Troubleshooting

- [X] I have read and followed the documentation AND the [troubleshooting guide](https://kyverno.io/docs/troubleshooting/).
- [X] I have searched other issues in this repository and mine is not recorded.

--- a/pkg/engine/context/deferred.go
+++ b/pkg/engine/context/deferred.go
@@ -168,6 +168,9 @@ func (d *deferredLoaders) setLevelAndIndex(level, index int) {
 }
 
 func (d *deferredLoaders) match(query string, level, index int) (*leveledLoader, int) {
+	if index > len(d.loaders) {
+		index = len(d.loaders)
+	}
 	for i := 0; i < index; i++ {
 		dl := d.loaders[i]
 		if dl.matched || dl.loader.HasLoaded() {

--- a/pkg/engine/context/deferred_test.go
+++ b/pkg/engine/context/deferred_test.go
@@ -412,3 +412,26 @@ func TestDeferredLoadMatchingError(t *testing.T) {
 	err := ctx.deferred.LoadMatching("value", len(ctx.jsonRawCheckpoints))
 	assert.ErrorContains(t, err, `failed to load data`)
 }
+
+type resetLoader struct {
+	ctx Interface
+}
+
+func (l *resetLoader) Name() string    { return "reset" }
+func (l *resetLoader) HasLoaded() bool { return false }
+func (l *resetLoader) LoadData() error {
+	l.ctx.Reset()
+	return nil
+}
+
+func TestDeferredResetDuringLoad(t *testing.T) {
+	ctx := newContext()
+	l1, _ := NewDeferredLoader("reset", &resetLoader{ctx: ctx}, logger)
+	l2, _ := NewDeferredLoader("other", &mockLoader{name: "other", ctx: ctx}, logger)
+	ctx.Checkpoint()
+	ctx.Checkpoint()
+	ctx.AddDeferredLoader(l1)
+	ctx.AddDeferredLoader(l2)
+	_, err := ctx.Query("[reset, other]")
+	assert.NilError(t, err)
+}

